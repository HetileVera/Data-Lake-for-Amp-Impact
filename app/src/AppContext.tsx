import React, { createContext, useContext } from "react";
import isEmpty from "lodash/isEmpty";
import throttle from "lodash/throttle";

import { AppStore, IAppStore } from "./models/AppStore";
import { IInstallation } from "./models/Installation";
import { isLocalDev, websiteUrlOrigin, assetWebsiteUrl, secretKey } from "./helpers/settings";
import { onSnapshot } from "mobx-state-tree";

// Create an instance of the root store (a.k.a appStore)
const appStore = AppStore.create({});

// Store state to local storage
const storeLocally = throttle(() => appStore.saveLocal(), 5000);

// Store state in S3 when applicable
const storeRemotely = throttle(() => appStore.saveRemote(), 30000);

// The initialization function for the root store
export function initializeStore() {
  if (!isEmpty(secretKey)) console.warn("Access key is bundled!");

  if (isLocalDev) {
    console.log(`isLocalDev '${isLocalDev}'`);
    console.log(`websiteUrlOrigin '${websiteUrlOrigin}'`);
    console.log(`assetWebsiteUrl '${assetWebsiteUrl}'`);

    if (window.location.host !== "localhost:3000") {
      console.warn("You should run the app on localhost!");

      throw new Error("App should run on localhost for development mode!");
    }
  }

  appStore.loadLocal();
  onSnapshot(appStore, (snapshot) => {
    storeLocally();
    storeRemotely();
  });
}

// Create the store context
export const StoreContext = createContext<IAppStore | undefined>(undefined);

// Create the store provider component
export const StoreProvider: React.FunctionComponent = ({ children }) => {
  return <StoreContext.Provider value={appStore}>{children}</StoreContext.Provider>;
};

// useStore hook
export function useStore(): IAppStore {
  const store = useContext(StoreContext);

  if (!store) {
    throw new Error("useStore() must be used within a StoreProvider");
  }

  return store;
}

// useInstallation hook
export function useInstallation(): IInstallation {
  const store = useStore();
  const installation = store.installation;

  if (!installation) {
    throw new Error("An installation instance was not created yet");
  }

  return installation;
}
