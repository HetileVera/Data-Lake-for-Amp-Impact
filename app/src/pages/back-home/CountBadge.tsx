import React from "react";
import { Badge } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useColorScheme } from "models/useColorScheme";
import { useMetadataStore } from "models/MetadataStore";

export const CountBadge = observer(() => {
  const { tone } = useColorScheme();
  const { isError, isReady, store } = useMetadataStore();

  if (isError || !isReady) return null;

  return (
    <Badge ml={3} p={1} pl={2} pr={2} bg={tone(100)} fontSize="small" borderRadius="full" color={tone(700)} fontWeight="bold">
      {store.selectedObjects.length}
    </Badge>
  );
});
