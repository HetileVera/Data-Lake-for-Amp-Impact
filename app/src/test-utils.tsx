import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";

import { theme } from "themes/orange";

export const CustomChakraProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <ChakraProvider theme={theme}>
      <div data-testid="test-root">{children}</div>
    </ChakraProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "wrapper">) => render(ui, { wrapper: CustomChakraProvider, ...options });

export * from "@testing-library/react";

export { customRender as render };
