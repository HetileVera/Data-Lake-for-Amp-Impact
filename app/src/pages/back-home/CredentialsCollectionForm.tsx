import React from "react";
import { Box, Button } from "@chakra-ui/react";
import { ArrowForwardIcon } from "@chakra-ui/icons";
import { observer } from "mobx-react";
import { SubmitHandler, useForm } from "react-hook-form";

import { useInstallation } from "AppContext";
import { useColorScheme } from "models/useColorScheme";
import { CredentialsValidationException } from "api/validate-credentials";

import { CredentialsForm, ICredentialsFormInput } from "pages/step1/CredentialsForm";
import { CredentialsError } from "pages/step1/CredentialsError";

interface ICredentialsCollectionForm {
  onCancel: () => void;
  onDone: () => void;
  buttonTitle?: string;
}

export const CredentialsCollectionForm = observer(({ onCancel, onDone, buttonTitle = "Resume" }: ICredentialsCollectionForm) => {
  const { colorScheme } = useColorScheme();
  const [error, setError] = React.useState<CredentialsValidationException | Error>();
  const installation = useInstallation();

  const onSubmit: SubmitHandler<ICredentialsFormInput> = async (values) => {
    setError(undefined);
    const accessKey = values.accessKeyId;
    const secretKey = values.secretAccessKey;
    const region = installation.region;

    try {
      await installation.connectToAwsStep.connectToAws(accessKey, secretKey, region);
      if (installation.isPostDeployment) installation.metadataStore.reset();

      return onDone();
    } catch (err: any) {
      console.log(err);
      setError(err);
    }
  };

  const {
    handleSubmit,
    register,
    control,
    formState: { errors, isSubmitting },
  } = useForm<ICredentialsFormInput>();

  const buttonProps: any = {};
  if (buttonTitle === "Resume") buttonProps.rightIcon = <ArrowForwardIcon />;

  return (
    <>
      <CredentialsError exception={error} />

      <Box pt={6}>
        {/* @ts-ignore */}
        <CredentialsForm {...{ register, errors, isSubmitting, control }} />
      </Box>

      <Box textAlign="right" w="full" mt={6} mb={0}>
        <Button colorScheme={colorScheme} size="md" disabled={isSubmitting} variant="ghost" onClick={() => onCancel()}>
          Cancel
        </Button>
        <Button
          colorScheme={colorScheme}
          size="md"
          loadingText="Connecting"
          ml={3}
          isLoading={isSubmitting}
          disabled={isSubmitting}
          onClick={handleSubmit(onSubmit)}
          {...buttonProps}
        >
          {buttonTitle}
        </Button>
      </Box>
    </>
  );
});
