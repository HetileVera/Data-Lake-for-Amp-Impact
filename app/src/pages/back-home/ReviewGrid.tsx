import React from "react";
import { Grid, GridItem } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useInstallation } from "AppContext";
import { isCompleted } from "models/Installation";
import { useColorScheme } from "models/useColorScheme";
import { awsRegionsMap } from "data/aws-regions";

export const ReviewGrid = observer(() => {
  const { tone } = useColorScheme();
  const installation = useInstallation();

  const commonProps = {
    w: "100%",
    h: "100%",
    borderRadius: "md",
    p: 0,
  };

  const headerProps = {
    ...commonProps,
    color: tone(700),
    fontWeight: "bold",
  };

  const contentProps = {
    ...commonProps,
    bg: tone(75),
    color: tone(800),
    p: 1,
    pl: 3,
  };

  return (
    <Grid templateColumns="0.5fr 1fr" gap={2} pb={6}>
      <GridItem {...headerProps}>AWS Account Id</GridItem>
      <GridItem {...contentProps}>{installation.accountId}</GridItem>

      <GridItem {...headerProps}>AWS Region</GridItem>
      <GridItem {...contentProps}>{awsRegionsMap[installation.region].label}</GridItem>

      {installation.appFlowConnectionName && (
        <>
          <GridItem {...headerProps}>Connection Name</GridItem>
          <GridItem {...contentProps}>{installation.appFlowConnectionName}</GridItem>
        </>
      )}
      {isCompleted(installation.connectToSalesforceStep) && (
        <>
          <GridItem {...headerProps}>Import Schedule</GridItem>
          <GridItem {...contentProps}>{installation.importOptionsStep.infoMessage}</GridItem>
        </>
      )}
    </Grid>
  );
});
