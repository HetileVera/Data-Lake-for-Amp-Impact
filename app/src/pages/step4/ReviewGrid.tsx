import React from "react";
import { Grid, GridItem } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useInstallation } from "AppContext";
import { niceNumber } from "helpers/utils";
import { awsRegionsMap } from "data/aws-regions";
import { useMetadataStore } from "models/MetadataStore";

export const ReviewGrid = observer(() => {
  const installation = useInstallation();
  const { store } = useMetadataStore();

  const commonProps = {
    w: "100%",
    h: "100%",
    borderRadius: "md",
    p: 3,
  };

  const headerProps = {
    ...commonProps,
    // bg: "pink.600",
    // color: "pink.50",
    color: "pink.700",
    fontWeight: "bold",
  };

  const contentProps = {
    ...commonProps,
    bg: "pink.75",
    color: "pink.800",
  };

  return (
    <Grid templateColumns="0.5fr 1fr" gap={2}>
      <GridItem {...headerProps}>AWS Region</GridItem>
      <GridItem {...contentProps}>{awsRegionsMap[installation.region].label}</GridItem>
      <GridItem {...headerProps}>Connection Name</GridItem>
      <GridItem {...contentProps}>{installation.appFlowConnectionName}</GridItem>
      <GridItem {...headerProps}>Import Schedule</GridItem>
      <GridItem {...contentProps}>{installation.importOptionsStep.infoMessage}</GridItem>
      <GridItem {...headerProps}>Data Model</GridItem>
      <GridItem {...contentProps}>We will import {niceNumber(store.selectedObjects.length)} objects</GridItem>
    </Grid>
  );
});
