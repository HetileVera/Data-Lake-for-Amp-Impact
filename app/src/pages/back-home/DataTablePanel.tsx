import React from "react";
import { Box, Heading } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useColorScheme } from "models/useColorScheme";
import { useMetadataStore } from "models/MetadataStore";

import { DataModelTablePreview } from "pages/step4/data-model-table-preview/DataModelTablePreview";

import { CountBadge } from "./CountBadge";
import { DataTableStatusInfo } from "./DataTableStatusInfo";

export const DataTablePanel = observer(() => {
  const { tone, colorScheme } = useColorScheme();
  const { store, isReady } = useMetadataStore();
  const selectedObjectsCount = store.selectedObjects.length;
  const isEmptyModel = selectedObjectsCount === 0;

  return (
    <>
      <Heading size="md" pt="0px" pb="0px" color={tone(600)} letterSpacing="-1px">
        Objects <CountBadge />
      </Heading>
      {isEmptyModel && isReady && (
        <Box bg={tone(100)} textAlign="center" p={4} mt={6} mb={4} fontSize="sm" color={tone(600)} borderRadius="md">
          No objects
        </Box>
      )}
      <DataTableStatusInfo />
      <DataModelTablePreview colorScheme={colorScheme} showCaption={false} />
    </>
  );
});
