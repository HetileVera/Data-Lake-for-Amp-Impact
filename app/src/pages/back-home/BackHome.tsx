import React from "react";
import { Box, Text, Button } from "@chakra-ui/react";
import { ArrowForwardIcon } from "@chakra-ui/icons";
import { observer } from "mobx-react";
import { useNavigate } from "react-router-dom";

import { useInstallation, useStore } from "AppContext";
import { Layout } from "components/Layout";
import { TimeAgo } from "components/TimeAgo";
import { isCompleted } from "models/Installation";
import { useColorScheme } from "models/useColorScheme";
import { AccessInformation } from "pages/step6/AccessInformation";
import { UsersPanel } from "pages/step6/users-panel/UsersPanel";
import { CloudwatchDashboardPanel } from "pages/step6/CloudwatchDashboardPanel";
import { SubscribeSNSPanel } from "pages/step6/SubscribeSNSPanel";
import { DataImportStatusPanel } from "pages/step5/data-import-status-panel/DataImportStatusPanel";

import { ReviewGrid } from "./ReviewGrid";
import { DataTablePanel } from "./DataTablePanel";
import { ManagePanel } from "./ManagePanel";
import { StartNewWarning } from "./StartNewWarning";
import { CredentialsCollectionForm } from "./CredentialsCollectionForm";

export const BackHome = observer(() => {
  const { tone, colorScheme } = useColorScheme();
  const navigate = useNavigate();
  const store = useStore();
  const installation = useInstallation();
  const [collectCredentials, setCollectCredentials] = React.useState(false);
  const [warning, setWarning] = React.useState(false);
  const completedProvisioning = isCompleted(installation.deploymentStep);
  const reviewCompleted = isCompleted(installation.reviewStep);
  const credentialsAvailable = !installation.credentials.empty;

  const gotoNextStep = () => {
    const nextStep = installation.nextStepNumber;
    navigate(`/steps/${nextStep}`);
  };

  const handleCollectCredentials = () => {
    if (credentialsAvailable) return gotoNextStep();
    setCollectCredentials(true);
  };

  const handleStartNew = () => {
    if (reviewCompleted) {
      setWarning(true);
    } else {
      store.newInstallation();
    }
  };

  React.useEffect(() => {
    window.scrollTo(0, 0);
  }, []);

  React.useEffect(() => {
    if (collectCredentials || warning) setTimeout(() => window.scrollBy(0, 300), 200);
  }, [collectCredentials, warning]);

  return (
    <Layout step={0} title="Data Lake Information">
      <Box position="relative" boxShadow="base" borderRadius="lg" bg={tone(50)} p={10} pb={6} color={tone(800)}>
        <Box fontSize="small" mt={1} mb={8} color={tone(700)}>
          <TimeAgo time={installation.startDate} /> &#8226; by {installation.startedBy}
        </Box>

        <ReviewGrid />

        {!completedProvisioning && !collectCredentials && !warning && (
          <>
            <Text pl={2} mt={6}>
              Click <b>Resume</b> to continue the steps of provisioning the data lake
              {!credentialsAvailable && <>, we will ask you for your AWS admin credentials so that we can connect to your AWS account your </>}. You
              can also choose to start a new data lake.
            </Text>
            <Box textAlign="right" w="full" mt={6} mb={0}>
              <Button colorScheme={colorScheme} size="md" variant="outline" mr={6} onClick={handleStartNew}>
                Start New
              </Button>
              <Button colorScheme={colorScheme} size="md" rightIcon={<ArrowForwardIcon />} onClick={handleCollectCredentials}>
                Resume
              </Button>
            </Box>
          </>
        )}
      </Box>

      {completedProvisioning && (
        <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={10}>
          <AccessInformation />
        </Box>
      )}

      {completedProvisioning && credentialsAvailable && (
        <>
          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={4} position="relative">
            <UsersPanel />
          </Box>

          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={6}>
            <DataImportStatusPanel />
          </Box>

          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={6}>
            <DataTablePanel />
          </Box>

          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={4} position="relative">
            <CloudwatchDashboardPanel region={installation.region} id={installation.id} />
          </Box>

          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={4} position="relative">
            <SubscribeSNSPanel />
          </Box>
        </>
      )}

      {completedProvisioning && !credentialsAvailable && <ManagePanel />}

      {warning && <StartNewWarning onCancel={() => setWarning(false)} onContinue={() => store.newInstallation()} />}

      {collectCredentials && (
        <CredentialsCollectionForm
          onCancel={() => setCollectCredentials(false)}
          onDone={() => {
            setCollectCredentials(false);
            gotoNextStep();
          }}
        />
      )}
      <Box h="50px">&nbsp;</Box>
    </Layout>
  );
});
