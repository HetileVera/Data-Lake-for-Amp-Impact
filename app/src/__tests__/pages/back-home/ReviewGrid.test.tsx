import { cleanup, screen } from "@testing-library/react";
import { create } from "react-test-renderer";

import { render, CustomChakraProvider } from "test-utils";

import { StoreProvider } from "AppContext";

import { ReviewGrid } from "pages/back-home/ReviewGrid";

import * as appContext from "AppContext";

describe("BackHome -> ReviewGrid component", () => {
  afterEach(cleanup);

  test("snapshot", () => {
    // @ts-ignore
    jest.spyOn(appContext, "useInstallation").mockImplementation(() => {
      return {
        accountId: "test-account-id",
        region: "us-east-1",
        appFlowConnectionName: "test-appflow-connection-name",
        connectToSalesforceStep: "test-step",
      };
    });

    const tree = create(
      <CustomChakraProvider>
        <StoreProvider>
          <ReviewGrid />
        </StoreProvider>
      </CustomChakraProvider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("render properly", () => {
    // @ts-ignore
    jest.spyOn(appContext, "useInstallation").mockImplementation(() => {
      return {
        accountId: "test-account-id",
        region: "us-east-1",
        appFlowConnectionName: "test-appflow-connection-name",
        connectToSalesforceStep: "test-step",
      };
    });

    render(
      <StoreProvider>
        <ReviewGrid />
      </StoreProvider>
    );

    expect(screen.getByText(/aws account id/i)).toBeInTheDocument();
    expect(screen.getByText(/test-account-id/i)).toBeInTheDocument();
    expect(screen.getByText(/aws region/i)).toBeInTheDocument();
    expect(screen.getByText(/connection name/i)).toBeInTheDocument();
    expect(screen.getByText(/test-appflow-connection-name/i)).toBeInTheDocument();
  });
});
