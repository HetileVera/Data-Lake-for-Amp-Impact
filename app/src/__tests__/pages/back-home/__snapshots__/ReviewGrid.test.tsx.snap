// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackHome -> ReviewGrid component snapshot 1`] = `
Array [
  <div
    className="css-iektmo"
  >
    <div
      className="css-1912uuf"
    >
      AWS Account Id
    </div>
    <div
      className="css-im2a0n"
    >
      test-account-id
    </div>
    <div
      className="css-1912uuf"
    >
      AWS Region
    </div>
    <div
      className="css-im2a0n"
    >
      US East (Northern Virginia) Region
    </div>
    <div
      className="css-1912uuf"
    >
      Connection Name
    </div>
    <div
      className="css-im2a0n"
    >
      test-appflow-connection-name
    </div>
  </div>,
  <span />,
]
`;
